servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ProxyConnectConnectionFactoryFilter.java

```diff
@@ -74,51 +68,63 @@ private ProxyFilter(final ConnectionFactory<ResolvedAddress, C> delegate) {
-                // HttpExecutionStrategy. Because we can't be sure if there is any blocking code in the connection
-                // filters we use the default strategy which should offload everything to be safe.
-                c.request(defaultStrategy(),
-                            reqRespFactory.connect(connectAddress).addHeader(CONTENT_LENGTH, ZERO))
-                 .flatMap(response -> {
-                if (SUCCESSFUL_2XX.contains(response.status())) {
-                    final Channel channel = ((NettyConnectionContext) c.connectionContext()).nettyChannel();
-                    final SingleSource.Processor<C, C> processor = newSingleProcessor();
+            return delegate().newConnection(resolvedAddress).flatMap(c -> {
+                try {
+                    // We currently only have access to a StreamingHttpRequester, which means we are forced to provide
+                    // an HttpExecutionStrategy. Because we can't be sure if there is any blocking code in the
+                    // connection filters we use the default strategy which should offload everything to be safe.
+                    return c.request(defaultStrategy(), c.connect(connectAddress).addHeader(CONTENT_LENGTH, ZERO))
+                            .flatMap(response -> handleConnectResponse(c, response))
+                            // Close recently created connection in case of any error while it connects to the proxy
+                            // or cancellation:
+                            .recoverWith(t -> c.closeAsync().concat(failed(t)))
+                            .whenCancel(() -> c.closeAsync().subscribe());
+                } catch (Throwable t) {
+                    return c.closeAsync().concat(failed(t));
+                }
+            });
+        }
+    }
 
-                    channel.pipeline().addLast(new ChannelInboundHandlerAdapter() {
-                        @Override
-                        public void userEventTriggered(final ChannelHandlerContext ctx, final Object evt) {
-                            if (evt instanceof SslHandshakeCompletionEvent) {
-                                SslHandshakeCompletionEvent event = (SslHandshakeCompletionEvent) evt;
-                                if (event.isSuccess()) {
-                                    processor.onSuccess(c);
-                                } else {
-                                    processor.onError(event.cause());
-                                }
-                            }
-                            ctx.fireUserEventTriggered(evt);
-                        }
-                    });
+    private Single<C> handleConnectResponse(final C connection, final StreamingHttpResponse response) {
+        try {
+            if (response.status().statusClass() != SUCCESSFUL_2XX) {
+                return response.payloadBodyAndTrailers().ignoreElements().concat(failed(
```

Comment:
this could lead to unnecessary resource (CPU, memory, network, etc...) consumption if we continuously read data that we don't care about (potential DOS vector). Can you clarify why this is necessary when we are hard-closing the connection?

Reply:
Good catch!
I thought about this while was working on this PR. But because it was deferred until after #990 is merged I forgot to ask about it. Fixed.